encoding default "null" failed in avro protocol

### What did you do?

1. run a changefeed with Avro protocol
2. run sql
```
create table t31(
    id int primary key,
    col varchar(255) DEFAULT "null"
);

insert into t31(id) values(2);
```

### What did you expect to see?

_No response_

### What did you see instead?
```
[ERROR] [schema_registry.go:335] ["GetCachedOrRegister: Could not make goavro codec"] [error="Record \"default.test.t31\" field \"col\": default value ought to encode using field schema: cannot encode binary bytes: expected: string; received: <nil>"]
[ERROR] [avro.go:102] ["AppendRowChangedEvent: avro encoding failed"] [error="[CDC:ErrAvroSchemaAPIError]schema manager API error: Record \"default.test.t31\" field \"col\": default value ought to encode using field schema: cannot encode binary bytes: expected: string; received: <nil>"] [errorVerbose="[CDC:ErrAvroSchemaAPIError]schema manager API error: Record \"default.test.t31\" field \"col\": default value ought to encode using field schema: cannot encode binary bytes: expected: string; received: <nil>\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:34\ngithub.com/pingcap/tiflow/pkg/sink/codec/avro.(*schemaManager).GetCachedOrRegister\n\tgithub.com/pingcap/tiflow/pkg/sink/codec/avro/schema_registry.go:336\ngithub.com/pingcap/tiflow/pkg/sink/codec/avro.(*BatchEncoder).avroEncode\n\tgithub.com/pingcap/tiflow/pkg/sink/codec/avro/avro.go:227\ngithub.com/pingcap/tiflow/pkg/sink/codec/avro.(*BatchEncoder).AppendRowChangedEvent\n\tgithub.com/pingcap/tiflow/pkg/sink/codec/avro/avro.go:100\ngithub.com/pingcap/tiflow/pkg/sink/codec.(*encoderGroup).runEncoder\n\tgithub.com/pingcap/tiflow/pkg/sink/codec/encoder_group.go:116\ngithub.com/pingcap/tiflow/pkg/sink/codec.(*encoderGroup).Run.func2\n\tgithub.com/pingcap/tiflow/pkg/sink/codec/encoder_group.go:95\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1598"]
```
### Versions of the cluster

TiCDC version (execute `cdc version`):

```console
v7.1.3
```

--- a/pkg/sink/codec/avro/avro.go
+++ b/pkg/sink/codec/avro/avro.go
@@ -560,8 +560,18 @@ func (a *BatchEncoder) columns2AvroSchema(tableName model.TableName, input avroE
 			}
 		} else {
 			if colx.GetFlag().IsNullable() {
+				// the string literal "null" must be coerced to a `nil`
+				// see https://github.com/linkedin/goavro/blob/5ec5a5ee7ec82e16e6e2b438d610e1cab2588393/record.go#L109-L114
 				// https://stackoverflow.com/questions/22938124/avro-field-default-values
+				defaultFirst := false
 				if defaultValue == nil {
+					defaultFirst = true
+				} else if s, ok := defaultValue.(string); ok && s == "null" {
+					defaultFirst = true
+				} else if b, ok := defaultValue.([]byte); ok && string(b) == "null" {
+					defaultFirst = true
+				}
+				if defaultFirst {
 					field["type"] = []interface{}{"null", avroType}
 				} else {
 					field["type"] = []interface{}{avroType, "null"}

--- a/tests/integration_tests/avro_basic/data/data.sql
+++ b/tests/integration_tests/avro_basic/data/data.sql
@@ -164,6 +164,15 @@ insert into t(c_tinyint, c_mediumint, c_int, c_bigint, a) values (4, 5, 6, 7, 8)
 alter table t modify c_mediumint varchar(10) null;
 insert into t(c_tinyint, c_mediumint, c_int, c_bigint, a) values (5, "234", 6, 7, 8);
 
+create table t1(
+    id int primary key,
+    c1 varchar(255) DEFAULT "null",
+    c2 varchar(255) DEFAULT "NULL",
+    c3 varchar(255) DEFAULT NULL
+);
+
+insert into t1(id) values(1);
+
 create table finish_mark
 (
     id int PRIMARY KEY

--- a/tests/integration_tests/avro_basic/data/data.sql
+++ b/tests/integration_tests/avro_basic/data/data.sql
@@ -166,13 +166,16 @@ insert into t(c_tinyint, c_mediumint, c_int, c_bigint, a) values (5, "234", 6, 7
 
 create table t1(
     id int primary key,
-    c1 varchar(255) DEFAULT "null",
-    c2 varchar(255) DEFAULT "NULL",
-    c3 varchar(255) DEFAULT NULL
+    c1 varchar(255) default "null",
+    c2 varchar(255) default "NULL",
+    c3 varchar(255) default null
 );
 
 insert into t1(id) values(1);
-update t1 set c1 = "null", c2 = "NULL", c3 = NULL where id = 1;
+update t1 set c1 = "null", c2 = "NULL", c3 = null where id = 1;
+alter table t1 add column col json not null;
+alter table t1 modify column col json default null;
+insert into t1(id) values(2);
 
 create table finish_mark
 (

--- a/tests/integration_tests/avro_basic/data/data.sql
+++ b/tests/integration_tests/avro_basic/data/data.sql
@@ -172,6 +172,7 @@ create table t1(
 );
 
 insert into t1(id) values(1);
+update t1 set c1 = "null", c2 = "NULL", c3 = NULL where id = 1;
 
 create table finish_mark
 (

--- a/tests/integration_tests/avro_basic/run.sh
+++ b/tests/integration_tests/avro_basic/run.sh
@@ -47,7 +47,7 @@ function run() {
 	SINK_URI="kafka://127.0.0.1:9092/$TOPIC_NAME?protocol=avro&enable-tidb-extension=true&avro-enable-watermark=true&avro-decimal-handling-mode=string&avro-bigint-unsigned-handling-mode=string"
 
 	run_cdc_cli changefeed create --start-ts=$start_ts --sink-uri="$SINK_URI" --config="$CUR/conf/changefeed.toml" --schema-registry=http://127.0.0.1:8088
-	cdc_kafka_consumer --upstream-uri $SINK_URI --downstream-uri="mysql://root@127.0.0.1:3306/?safe-mode=true&batch-dml-enable=false" --upstream-tidb-dsn="root@tcp(${UP_TIDB_HOST}:${UP_TIDB_PORT})/?" --schema-registry-uri=http://127.0.0.1:8088 --config="$CUR/conf/changefeed.toml" 2>&1 &
+	run_kafka_consumer $WORK_DIR $SINK_URI "$CUR/conf/changefeed.toml" http://127.0.0.1:8088
 
 	run_sql_file $CUR/data/data.sql ${UP_TIDB_HOST} ${UP_TIDB_PORT}
 

